File: learn/sorting/merge-sort/merge_sort.go


type MergeSort struct{}

func (s *MergeSort) Sort(arr []int) {
	if len(arr) > 0 {
		s.mergeSort(arr, 0, len(arr)-1)
	}
}

func (s *MergeSort) mergeSort(arr []int, left int, right int) {
	if left < right {
		mid := left + (right-left)/2
		s.mergeSort(arr, left, mid)
		s.mergeSort(arr, mid+1, right)
		s.merge(arr, left, mid, right)
	}
}

func (s *MergeSort) merge(arr []int, left int, mid int, right int) {
	n1 := mid - left + 1
	n2 := right - mid

	L := make([]int, n1)
	R := make([]int, n2)

	for i := 0; i < n1; i++ {
		L[i] = arr[left+i]
	}
	for j := 0; j < n2; j++ {
		R[j] = arr[mid+1+j]
	}

	i, j, k := 0, 0, left
	for i < n1 && j < n2 {
		if L[i] <= R[j] {
			arr[k] = L[i]
			i++
		} else {
			arr[k] = R[j]
			j++
		}
		k++
	}

	for i < n1 {
		arr[k] = L[i]
		i++
		k++
	}
	for j < n2 {
		arr[k] = R[j]
		j++
		k++
	}
}

File: learn/sorting/merge-sort/merge_sort.rs
pub struct MergeSort;

impl MergeSort {
    pub fn sort(arr: &mut [i32]) {
        let n = arr.len();
        if n > 0 {
            Self::merge_sort(arr, 0, n - 1);
        }
    }

    fn merge_sort(arr: &mut [i32], left: usize, right: usize) {
        if left < right {
            let mid = left + (right - left) / 2;
            Self::merge_sort(arr, left, mid);
            Self::merge_sort(arr, mid + 1, right);
            Self::merge(arr, left, mid, right);
        }
    }

    fn merge(arr: &mut [i32], left: usize, mid: usize, right: usize) {
        let n1 = mid - left + 1;
        let n2 = right - mid;

        let mut l = Vec::with_capacity(n1);
        let mut r = Vec::with_capacity(n2);

        for i in 0..n1 {
            l.push(arr[left + i]);
        }
        for j in 0..n2 {
            r.push(arr[mid + 1 + j]);
        }

        let mut i = 0;
        let mut j = 0;
        let mut k = left;

        while i < n1 && j < n2 {
            if l[i] <= r[j] {
                arr[k] = l[i];
                i += 1;
            } else {
                arr[k] = r[j];
                j += 1;
            }
            k += 1;
        }

        while i < n1 {
            arr[k] = l[i];
            i += 1;
            k += 1;
        }
        while j < n2 {
            arr[k] = r[j];
            j += 1;
            k += 1;
        }
    }
}
